### PR TITLE
Added support for incremental solving for the Boolector SMT backend

### PR DESCRIPTION
**Pull request by lucasccordeiro**
The main goal of this PR is to add support for incremental solving for the Boolector SMT backend. It also involves some refactoring of the SMT backend to pass a const reference of the option `optionst` object to the backend and check the options from it, as discussed in this PR https://github.com/esbmc/esbmc/pull/354.

### CODE REVIEW

File: regression/CMakeLists.txt [hunk 1]
```diff
@@ -26,12 +26,12 @@ endfunction()
...
 # NOTE: In order to make the best of the concurrency set sort the tests from the slowest to fastest.
 if(APPLE)
-    set(REGRESSIONS esbmc-unix esbmc-unix2 esbmc cbmc cstd llvm floats floats-regression k-induction k-induction-parallel nonz3 bitwuzla)
+    set(REGRESSIONS esbmc-unix esbmc-unix2 esbmc cbmc cstd llvm floats floats-regression k-induction k-induction-parallel nonz3 bitwuzla incremental)
 elseif(WIN32)
     # FUTURE: Add floats-regression esbmc-cpp/cpp
     set(REGRESSIONS esbmc cbmc cstd llvm floats  k-induction )
 else()
-    set(REGRESSIONS esbmc-unix esbmc-unix2 esbmc cbmc cstd llvm floats floats-regression k-induction esbmc-cpp/cpp csmith k-induction-parallel nonz3 bitwuzla)
+    set(REGRESSIONS esbmc-unix esbmc-unix2 esbmc cbmc cstd llvm floats floats-regression k-induction esbmc-cpp/cpp csmith k-induction-parallel nonz3 bitwuzla incremental)
```
Comment (rafaelsamenezes) [inline regression/CMakeLists.txt]: Here incremental means incremental SMT solving, right? Could you rename? Incremental will probably be confused with incremental-bmc for newer test cases.

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]
Just a small nitpick.

File: regression/CMakeLists.txt (hunk 1 above)
Comment (lucasccordeiro) [reply regression/CMakeLists.txt]: Yes. I have renamed the folder to `incremental-smt`.

**Review by mikhailramalho — APPROVED** [review]
LGTM.

Please, just double-check if you don't need to set flags in the other backends as well.

**Pull request merged**